compiler/rustc_ast_lowering/src/local.rs

```diff
@@ -0,0 +1,68 @@
+use super::LoweringContext;
+
+use rustc_ast::*;
+use rustc_data_structures::fx::FxIndexMap;
+use rustc_data_structures::thin_vec::ThinVec;
+use rustc_hir as hir;
+use rustc_session::parse::feature_err;
+use rustc_span::symbol::Ident;
+use rustc_span::{sym, DUMMY_SP};
+
+impl<'hir> LoweringContext<'_, 'hir> {
+    /// Before: `let Some([ref x, ref y]) = z else { return }`
+    /// After: `let (ref x, ref y) = match z { Some([x, y]) => (x, y), _ => return }`
```

Comment:
I am not convinced this is the best desugaring for this feature. I would rather see something like:
Before: `let Some([ref x, ref y]) = z else { return }; fun(x, y)`
After: `match z { Some([ref x, ref y]) => { fun(x, y) }, _ => { return }}`.
This would avoid the need for `let_else_bindings` and would avoid duplicating the bindings three times.

Reply:
The bindings need to be in scope for later statements. I don't think that is possible with your example.

Reply:
Wait...are you suggesting to put the remainder of the block inside the `match`?

Reply:
Yes :) 
However, this suggestion will require to be careful with lint attributes on the let-else statement.

Reply:
Hmm right because the attribute should only apply to the then-pattern and the else-block. Should I duplicate the attributes? Or put some workaround in the visitor?

Reply:
I think patterns do not have attributes so I would just put them on the else-block.

I found another problem - where would the type annotation go? Is `hir::ExprKind::Type` on the initializer an option, or no because type ascription is unstable? Or I suppose I could desugar into two statements? `let x: ty = ..; match x { .. }`

Reply:
> Is `hir::ExprKind::Type` on the initializer an option?

No because it would not allow coercion.

Reply:
Is this change to the desugaring documented/discussed anywhere? I am not complaining, but it actually goes beyond the desugaring proposed in the RFC in that it allows a `ref` binding to a non-copy value, as exercised in the run-pass test. So do we need to note this somewhere (besides the test) to indicate that it's an intended addition?

Reply:
@durka IMO the desugaring is an implementation detail and therefore technically not within the scope of what is specified by the RFC. Though the desugaring in the RFC works as an illustration of the expected behavior of let-else. I don't see anything that prohibits `ref` in the pattern. In user code, I would expect a `ref` to be allowed in exactly the same manner as a normal `let` or `if let`.